src/Common/src/Interop/FreeBSD/Interop.Process.cs

```diff
@@ -0,0 +1,445 @@
+            public rusage ki_rusage;                    /* process rusage statistics */
+            /* XXX - most fields in ki_rusage_ch are not (yet) filled in */
+            private rusage ki_rusage_ch;                /* rusage of children processes */
+            private void* ki_pcb;                       /* kernel virtual addr of pcb */
+            private void* ki_kstack;                    /* kernel virtual addr of stack */
+            private void* ki_udata;                     /* User convenience pointer */
+            public void* ki_tdaddr;                     /* address of thread */
+
+            private fixed long ki_spareptrs[KI_NSPARE_PTR];     /* spare room for growth */
+            private fixed long ki_sparelongs[KI_NSPARE_LONG];   /* spare room for growth */
+            private long ki_sflag;                              /* PS_* flags */
+            private long ki_tdflags;                            /* XXXKSE kthread flag */
+        }
+
+        /// <summary>
+        /// Queries the OS for the list of all running processes and returns the PID for each
+        /// </summary>
+        /// <returns>Returns a list of PIDs corresponding to all running processes</returns>
+        internal static unsafe int[] proc_listallpids()
+        {
+            int numProcesses=0;
+            int[] pids;
+            kinfo_proc * entries = null;
+            int idx;
+
+            try
+            {
+                entries = getProcInfo(0, false, out numProcesses);
+                if (entries == null || numProcesses <= 0)
+                {
+                    throw new Win32Exception(SR.CantGetAllPids);
+                }
+
+                Span<kinfo_proc>  list = new Span<kinfo_proc>(entries, numProcesses);
+                pids = new int[numProcesses];
+                idx=0;
+                // walk through process list and skip kernel threads
+                for (int i=0; i < list.Length; i++)
+                {
+                    if (list[i].ki_ppid == 0)
+                    {
+                        // skip kernel threads
+                        numProcesses-=1;
+                    }
+                    else
+                    {
+                        pids[idx] = list[i].ki_pid;
+                        idx += 1;
+                    }
+                }
+                // Remove extra elements
+                Array.Resize<int>(ref pids, numProcesses);
+            }
+            finally
+            {
+                Marshal.FreeHGlobal((IntPtr)entries);
+            }
+            return pids;
+        }
+
+
+        /// <summary>
+        /// Gets executable name for process given it's PID
+        /// </summary>
+        /// <param name="pid">The PID of the process</param>
+        public static unsafe string getProcPath(int pid)
+        {
+            Span<int> name = stackalloc int[4];
+            byte* pBuffer = stackalloc byte[PATH_MAX];
+            int bytesLength = PATH_MAX;
```

Comment:
This sysctl(3) does not allocate array, but writes to a buffer of desired length.

In the NetBSD case `PATH_MAX` (1024 bytes) should be used in the current implementation, FreeBSD is more tolerant for shorter buffers.

Reply:
Ideally there should be a fallback for parsing `argv[0]` as this call might not be reliable. It also gives wrong results on both BSDs for hardlinks (internally it uses a vnode to path translator), there are bugs with null-mounts etc..

Reply:
@krytarowski the buffer allocation based on the necessary size happens in the Interop.Sys.SysCtl implementation if null is passed as the buffer and length of zero is passed:
```C#
            if (value == null && len == 0) 
            { 
               // do one try to see how much data we need 
                ret = SysctlByName(name, pBuffer, &bytesLength); 
```

```

Reply:
I've verified that this trick works on NetBSD for this call, so it's fine too.

Reply:
Did we reached consensus here? The intentions to avoid allocation for temporary buffer. 
It is my understanding that OS cannot have longer names so PTH_MAX is proper upper bound. I'd be happy to change to dynamic allocation. This part is probably not performance critical anyway.

Reply:
@janvorli suggested a good enough solution for 98% use-cases. Remaining 2% are kernel bugs here and there and they should be addressed by kernel programmers. (No need to list the constrains here).

Reply:
I would be surprised if the OS had 1024 character limit on path length. At least on Linux, this limit is file system specific and there are file systems with much higher limit.

Reply:
Modern code should use pathconf(3) and prompt per-filesystem specific maximum paths.

```
     _PC_PATH_MAX
             The maximum number of bytes in a pathname, including the
             terminating null character.
```

However it's difficult to get rid of the old hardcoded values and usually they are large enough to focus on more important areas. Also POSIX mandates some hardcoded values... (not sure about this one).